File: src/main/java/io/github/addoncommunity/galactifun/api/universe/UniversalObject.java
package io.github.addoncommunity.galactifun.api.universe;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;

import javax.annotation.Nonnull;

import lombok.Getter;

import org.bukkit.inventory.ItemStack;

import io.github.addoncommunity.galactifun.api.universe.attributes.Orbit;
import io.github.addoncommunity.galactifun.api.universe.types.UniversalType;
import io.github.thebusybiscuit.slimefun4.libraries.dough.items.CustomItemStack;
import io.github.thebusybiscuit.slimefun4.utils.ChatUtils;

/**
 * Any object in the universe
 *
 * @author Mooy1
 */
public abstract class UniversalObject {

    private final List<UniversalObject> orbiters = new ArrayList<>();
    @Getter
    private final String name;
    @Getter
    protected final String id;
    @Getter
    private final ItemStack item;
    @Getter
    private final UniversalObject orbiting;
    @Getter
    private final Orbit orbit;
    protected final int orbitLevel;

    UniversalObject(@Nonnull String name, @Nonnull UniversalType type, @Nonnull Orbit orbit,
                    @Nonnull UniversalObject orbiting, @Nonnull ItemStack baseItem) {
        this.name = ChatUtils.removeColorCodes(name);
        this.item = CustomItemStack.create(baseItem, name, "&7Type: " + type.description());
        this.orbiting = orbiting;
        this.orbit = orbit;
        this.orbitLevel = orbiting.orbitLevel + 1;
        this.id = this.name.toLowerCase(Locale.ROOT).replace(' ', '_');
        orbiting.orbiters.add(this);
    }

    /**
     * Constructor for the universe
     */
    UniversalObject(String name) {
        this.name = ChatUtils.removeColorCodes(name);
        this.id = this.name.toLowerCase(Locale.ROOT).replace(' ', '_');
        this.item = null;
        this.orbiting = null;
        this.orbit = null;
        this.orbitLevel = 0;
    }

    /**
     * Gets the distance in light years between 2 objects
     */
    public double distanceTo(@Nonnull UniversalObject other) {
        if (this.orbiting == other.orbiting) {
            double thisDist = this.orbit.currentDistance();
            double otherDist = other.orbit.currentDistance();
            double cosAngle = Math.cos(this.orbit.position() - other.orbit.position());
            return Math.sqrt(thisDist * thisDist + otherDist * otherDist - (2 * thisDist * otherDist * cosAngle));
        }
        if (this.orbiting == null || this.orbitLevel < other.orbitLevel) {
            return other.orbit.currentDistance() + distanceTo(other.orbiting);
        }
        return this.orbit.currentDistance() + this.orbiting.distanceTo(other);
    }

    @Nonnull
    public List<UniversalObject> orbiters() {
        return Collections.unmodifiableList(this.orbiters);
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

}

File: src/main/kotlin/io/github/addoncommunity/galactifun/api/objects/UniversalObject.kt
package io.github.addoncommunity.galactifun.api.objects

import io.github.addoncommunity.galactifun.api.objects.properties.Orbit
import io.github.bakedlibs.dough.items.CustomItemStack
import io.github.thebusybiscuit.slimefun4.utils.ChatUtils
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import kotlin.math.cos
import kotlin.math.sqrt

abstract class UniversalObject private constructor(
    name: String,
    baseItem: ItemStack,
    // Special case for The Universe
    private val _orbiting: UniversalObject?,
    val orbit: Orbit
) {

    val name = ChatUtils.removeColorCodes(name)
    val id = this.name.lowercase().replace(' ', '_')

    val item = CustomItemStack(baseItem, name)

    val orbiting: UniversalObject
        get() = _orbiting ?: error("The Universe does not orbit anything")

    val orbitLevel: Int = _orbiting?.let { it.orbitLevel + 1 } ?: 0

    private val _orbiters = mutableListOf<UniversalObject>()
    val orbiters: List<UniversalObject>
        get() = _orbiters.toList()

    internal constructor() : this("The Universe", ItemStack(Material.NETHER_STAR), null, Orbit(0.0))

    protected constructor(name: String, baseItem: ItemStack, orbiting: UniversalObject, orbit: Orbit) : this(
        name,
        baseItem,
        // Force it to call the private constructor
        orbiting as UniversalObject?,
        orbit
    )

    protected constructor(name: String, baseItem: Material, orbiting: UniversalObject, orbit: Orbit) : this(
        name,
        ItemStack(baseItem),
        orbiting,
        orbit
    )

    fun addOrbiter(orbiter: UniversalObject) {
        _orbiters.add(orbiter)
    }

    open fun distanceTo(other: UniversalObject): Double {
        if (_orbiting == null || orbitLevel < other.orbitLevel) {
            return other.orbit.distance + distanceTo(other.orbiting)
        }
        if (orbiting == other.orbiting) {
            val thisDist = orbit.distance
            val otherDist = other.orbit.distance
            val cosAngle = cos(orbit.position - other.orbit.position)
            return sqrt(thisDist * thisDist + otherDist * otherDist - 2 * thisDist * otherDist * cosAngle)
        }
        return orbit.distance + orbiting.distanceTo(other)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is UniversalObject) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()
}